src/main/scala/is/hail/expr/ir/MatrixIR.scala

```diff
@@ -152,9 +152,39 @@ object MatrixIR {
 abstract sealed class MatrixIR extends BaseIR {
   def typ: MatrixType
 
-  def partitionCounts: Option[IndexedSeq[Long]] = None
+  def rowCount: Long = partitionCounts.sum
 
-  def columnCount: Option[Int] = None
+  var _columnCount: Option[Int] = None
+
+  def columnCount: Long = {
+    if (fastColumnCount.isEmpty)
+      _columnCount = Some(MatrixColsTable(this).partitionCounts.sum.toInt)
+    _columnCount.get
+  }
+
+  def fastColumnCount: Option[Int] = {
+    if (_columnCount.isEmpty)
+      _columnCount = _fastColumnCount
+    _columnCount
+  }
+
+  def _fastColumnCount: Option[Int] = None
+
+  var _partitionCounts: Option[IndexedSeq[Long]] = None
+
+  def partitionCounts: IndexedSeq[Long] = {
+    if (fastPartitionCounts.isEmpty)
+      _partitionCounts = Some(MatrixRowsTable(this).partitionCounts)
+    _partitionCounts.get
+  }
+
+  def fastPartitionCounts: Option[IndexedSeq[Long]] = {
+    if (_partitionCounts.isEmpty)
+      _partitionCounts = _fastPartitionCounts
```

Comment:
we should push this back down the IR when we execute, but don't need to do that in the same PR.